ethereum/core/src/test/java/tech/pegasys/pantheon/ethereum/worldstate/PrunerTest.java

```diff
@@ -0,0 +1,173 @@
+    final BlockchainStorage blockchainStorage =
+        new KeyValueStoragePrefixedKeyBlockchainStorage(
+            new InMemoryKeyValueStorage(), new MainnetBlockHeaderFunctions());
+    final DefaultMutableBlockchain blockchain =
+        new DefaultMutableBlockchain(genesisBlock, blockchainStorage, metricsSystem);
+
+    final Pruner pruner = new Pruner(markSweepPruner, blockchain, mockExecutorService, 1, 2);
+    pruner.start();
+
+    final Hash markBlockStateRootHash =
+        appendBlockWithParent(blockchain, genesisBlock).getHeader().getStateRoot();
+    verify(markSweepPruner, never()).mark(markBlockStateRootHash);
+    verify(markSweepPruner, never()).sweep();
+
+    appendBlockWithParent(blockchain, blockchain.getChainHeadBlock());
+    verify(markSweepPruner).mark(markBlockStateRootHash);
+    verify(markSweepPruner, never()).sweep();
+
+    appendBlockWithParent(blockchain, blockchain.getChainHeadBlock());
+    verify(markSweepPruner).sweep();
+    pruner.stop();
+  }
+
+  @Test
+  public void abortsPruningWhenFullyMarkedBlockNoLongerOnCanonicalChain()
+      throws InterruptedException {
+    final BlockchainStorage blockchainStorage =
+        new KeyValueStoragePrefixedKeyBlockchainStorage(
+            new InMemoryKeyValueStorage(), new MainnetBlockHeaderFunctions());
+    final DefaultMutableBlockchain blockchain =
+        new DefaultMutableBlockchain(genesisBlock, blockchainStorage, metricsSystem);
+
+    // start pruner so it can start handling block added events
+    final Pruner pruner = new Pruner(markSweepPruner, blockchain, mockExecutorService, 0, 1);
+    pruner.start();
+
+    /*
+     Set up pre-marking state:
+      O <---- marking of the this block's parent will begin when this block is added
+      |
+      |  O <- this is a fork as of now (non-canonical)
+      O  | <- this is the initially canonical block that will be marked
+       \/
+       O <--- the common ancestor when the reorg happens
+    */
+    final Block initiallyCanonicalBlock = appendBlockWithParent(blockchain, genesisBlock);
+    appendBlockWithParent(blockchain, initiallyCanonicalBlock);
+    final Block forkBlock = appendBlockWithParent(blockchain, genesisBlock);
+    /*
+      Cause reorg:
+     Set up pre-marking state:
+      O
+      |  O <---- this block causes a reorg; this branch becomes canonical
+      |  O <---- which means that state here is referring to nodes from the common ancestor block,
+      O  | <- because this was block at which marking began
+       \/
+       O
+    */
+    appendBlockWithParent(blockchain, forkBlock);
+    verify(markSweepPruner).mark(initiallyCanonicalBlock.getHeader().getStateRoot());
+    verify(markSweepPruner, never()).sweep();
+    pruner.stop();
+  }
+
+  @Test
+  public void shouldRejectInvalidArguments() {
+    final Blockchain mockchain = mock(Blockchain.class);
+    assertThatThrownBy(() -> new Pruner(markSweepPruner, mockchain, mockExecutorService, -1, -2))
```

Comment:
nit: in a previous commit you covered multiple cases where only one arg was illegal (i.e. 0, -1 and what should have been -1, 0). I felt that test had more coverage.

Reply:
Yes, meredith and I realized that the case that I thought was illegal is actually possible, so we relaxed the cases where the arguments were illegal.